api/src/accounts.rs

```diff
@@ -193,25 +208,67 @@ pub struct Account {
 }
 
 impl Account {
-    /// Creates a new account struct and determines the current ledger info, and determines the
-    /// ledger version to query
     pub fn new(
         context: Arc<Context>,
         address: Address,
         requested_ledger_version: Option<U64>,
         start: Option<StateKey>,
         limit: Option<u16>,
+        no_internal_indexer_required: bool,
     ) -> Result<Self, BasicErrorWith404> {
-        // Use the latest ledger version, or the requested associated version
-        let (latest_ledger_info, requested_ledger_version) = context
-            .get_latest_ledger_info_and_verify_lookup_version(
-                requested_ledger_version.map(|inner| inner.0),
-            )?;
+        let sharding_enabled = context
+            .node_config
+            .storage
+            .rocksdb_configs
+            .enable_storage_sharding;
+        let internal_indexer_enabled = context
+            .node_config
+            .indexer_db_config
+            .is_internal_indexer_db_enabled();
+
+        let (latest_ledger_info, requested_version) =
+            if sharding_enabled && !no_internal_indexer_required {
+                if !internal_indexer_enabled {
+                    return Err(BasicErrorWith404::internal_with_code_no_info(
+                        "Internal indexer is not enabled",
+                        AptosErrorCode::InternalError,
+                    ));
+                }
+                let (latest_internal_indexer_ledger_info, latest_internal_indexer_ledger_version) =
+                    context.get_latest_internal_indexer_ledger_version_and_info()?;
+                if let Some(version) = requested_ledger_version {
+                    let request_ledger_version = Version::from(version);
+                    if latest_internal_indexer_ledger_version < request_ledger_version {
+                        return Err(version_not_found(
+                            request_ledger_version,
+                            &latest_internal_indexer_ledger_info,
+                        ));
+                    } else if request_ledger_version
+                        < latest_internal_indexer_ledger_info.oldest_ledger_version.0
+                    {
+                        return Err(version_pruned(
+                            request_ledger_version,
+                            &latest_internal_indexer_ledger_info,
+                        ));
+                    }
+                    (latest_internal_indexer_ledger_info, request_ledger_version)
+                } else {
+                    (
+                        latest_internal_indexer_ledger_info,
+                        latest_internal_indexer_ledger_version,
+                    )
+                }
```

Comment:
shouldn't these be part of the context like `.get_latest_ledger_info_and_verify_lookup_version()`  in the `else` branch, for parity? i.e. maybe `.get_latest_indexer_ledger_info_and_verify_lookup_version()`

actually, I think the entirety of this `if-else` should belong to the context, maybe `.get_latest_ledger_info_and_verify_lookup_version(requred_version_opt, require_indices)`

might also need `context.ensure_state_indices_available()`, `context.ensure_event_indices_available()` etc

Reply:
refactored. I think the api at internal indexer level checks if corresponding config is enabled and will show corresponding error message